File: swift/exercises.swift
import Foundation

struct NegativeAmountError: Error {}
struct NoSuchFileError: Error {}

func change(_ amount: Int) -> Result<[Int:Int], NegativeAmountError> {
    if amount < 0 {
        return .failure(NegativeAmountError())
    }
    var (counts, remaining) = ([Int:Int](), amount)
    for denomination in [25, 10, 5, 1] {
        (counts[denomination], remaining) = 
            remaining.quotientAndRemainder(dividingBy: denomination)
    }
    return .success(counts)
}

func firstThenLowerCase(of: Array<String>, satisfying: (String) -> Bool) -> String?{
    var first: String?
    var firstFound: Bool = false

    for string in of {
        if (satisfying(string) && !firstFound) {
            first = string
            firstFound = true
        }
    }

    if (firstFound) {
        return first?.lowercased()
    }
    return nil
}

class ChainableString {
    var phrase: String

    init(_ phrase: String) {
        self.phrase = phrase
    }

    func and(_ stringSuffix: String) -> ChainableString {
        return ChainableString(phrase + " " + stringSuffix)
    }
}

func say(_ string: String = "") -> ChainableString {
    return ChainableString(string)
}

enum FileError: Error {
    case fileNotFound(String)
}

func meaningfulLineCount(_ path: String) async -> Result<Int, Error> {
    guard let contents = try? String(contentsOfFile: path) else {
        return .failure(NoSuchFileError())
    }

    let lines = contents.replacingOccurrences(of: "\r\n", with: "\n").split(separator: "\n")

    var lineCount: Int = 0
    for singleLine in lines {
        var firstChar = "#"
        var firstCharFound = false
        
        let chars = Array(singleLine)
        for char in chars {
            // print("\nChecking char: \(String(describing: char.asciiValue))")
            if (char.asciiValue != 13 && char.asciiValue != 0 && char.asciiValue != 32 && char.asciiValue != 9 && !firstCharFound) {
                firstChar = String(char)
                firstCharFound = true
            }
        }
        
        if (firstChar != "#" && firstCharFound) {
            lineCount += 1
        }
    }

    return .success(lineCount)
}

// Write your Quaternion struct here
// make copy of list when constructing quaternion
// internal list can be mutable
// return immutable list for quaternion
struct Quaternion: CustomStringConvertible {
    let a: Double
    let b: Double
    let c: Double
    let d: Double

    init(a: Double = 0, b: Double = 0, c: Double = 0, d: Double = 0) {
        self.a = a
        self.b = b
        self.c = c
        self.d = d    
    }
    
    var coefficients: [Double] { return [self.a, self.b, self.c, self.d] }
    var conjugate: Quaternion { return Quaternion(a: self.a, b: -self.b, c: -self.c, d: -self.d) }
    
    var description: String {
        var quatString: String = ""
        
        if (self.a != 0.0) { quatString = "\(self.a)" }
        
        if (self.b == 1.0 && quatString.length == 0) { quatString = "\(quatString)i" }
        else if (self.b == 1.0) { quatString = "\(quatString)+i" }
        else if (self.b == -1.0) { quatString = "\(quatString)-i" }
        else if (self.b > 0.0) { quatString = "\(quatString)+\(self.b)i" }
        else if (self.b < 0.0) { quatString = "\(quatString)\(self.b)i" }
        
        if (self.c == 1.0 && quatString.length == 0) { quatString = "\(quatString)j" }
        else if (self.c == 1.0) { quatString = "\(quatString)+j" }
        else if (self.c == -1.0) { quatString = "\(quatString)-j" }
        else if (self.c > 0.0) { quatString = "\(quatString)+\(self.c)j" }
        else if (self.c < 0.0) { quatString = "\(quatString)\(self.c)j" }
        
        if (self.d == 1.0 && quatString.length == 0) { quatString = "\(quatString)k" }
        else if (self.d == 1.0) { quatString = "\(quatString)+k" }
        else if (self.d == -1.0) { quatString = "\(quatString)-k" }
        else if (self.d > 0.0) { quatString = "\(quatString)+\(self.d)k" }
        else if (self.d < 0.0) { quatString = "\(quatString)\(self.d)k" }
        
        if (quatString.length == 0) { quatString = "0" }
        
        return quatString
    }

    static func == (left: Quaternion, right: Quaternion) -> Bool {
        return left.a == right.a && left.b == right.b && left.c == right.c && left.d == right.d
    }

    static func + (left: Quaternion, right: Quaternion) -> Quaternion {
        return Quaternion(a: left.a + right.a, b: left.b + right.b, c: left.c + right.c, d: left.d + right.d)
    }

    static func * (left: Quaternion, right: Quaternion) -> Quaternion {
        return Quaternion(a: (left.a * right.a) - (left.b * right.b) - (left.c * right.c) - (left.d * right.d),
                          b: (left.a * right.b) + (left.b * right.a) + (left.c * right.d) - (left.d * right.c),
                          c: (left.a * right.c) - (left.b * right.d) + (left.c * right.a) + (left.d * right.b),
                          d: (left.a * right.d) + (left.b * right.c) - (left.c * right.b) + (left.d * right.a))
    }

    static let ZERO = Quaternion(a: 0, b: 0, c: 0, d: 0)
    static let I = Quaternion(a: 0, b: 1, c: 0, d: 0)
    static let K = Quaternion(a: 0, b: 0, c: 0, d: 1)
    static let J = Quaternion(a: 0, b: 0, c: 1, d: 0)
}

indirect enum BinarySearchTree: CustomStringConvertible {
    case empty
    case nodeTree(tree: Array<String>, count: Int)

    var description: String {
        switch self {
            case .empty:
                return "()"
            case .nodeTree(let tree, _):
                var outputString: String = tree[1]
                if (tree.count > 2 && tree[2] != "None") {
                    outputString = descriptionHelper(tree: tree, outputString, 2)
                }
                if (tree.count > 3 && tree[3] != "None") {
                    outputString = descriptionHelper(tree: tree, outputString, 3)
                }
                return "(" + outputString + ")"
        }
    }

    func descriptionHelper(tree: Array<String>, _ c: String, _ i: Int) -> String {
        var parentStringIndex = c.range(of: tree[parent(i)])!.lowerBound
        var outputString: String = ""

        let insertingString = "(" + tree[i] + ")"
        if (i == leftChild(parent(i))) {
            outputString = c.prefix(upTo: parentStringIndex) + insertingString + c.suffix(from: parentStringIndex)
        } else {
            parentStringIndex = c.index(parentStringIndex, offsetBy: 1)
            outputString = c.prefix(upTo: parentStringIndex) + insertingString + c.suffix(from: parentStringIndex)
        }

        if (leftChild(i) < tree.count && tree[leftChild(i)] != "None") {
            outputString = descriptionHelper(tree: tree, outputString, leftChild(i))
        }
        if (rightChild(i) < tree.count && tree[rightChild(i)] != "None") {
            outputString = descriptionHelper(tree: tree, outputString, rightChild(i))
        }

        return outputString
    }

    var size: Int {
        switch self {
            case .empty:
                return 0
            case .nodeTree(_, let count):
                return count
        }
    }

    func parent(_ i: Int) -> Int { return i/2 }
    func leftChild(_ i: Int) -> Int { return i*2 }
    func rightChild(_ i: Int) -> Int { return i*2 + 1 }

    func insert(_ c: String) -> BinarySearchTree {
        switch self {
            case .empty:
                return BinarySearchTree.nodeTree(tree: ["None", c], count: 1)
            case .nodeTree(let tree, let count):
                var newTree: Array<String> = tree.map { $0 }
                var insertIndexFound: Bool = false
                var currentIndex: Int = 1

                while (!insertIndexFound) {
                    if (Character(newTree[currentIndex]).asciiValue ?? UInt8(0) > Character(c).asciiValue ?? UInt8(0)) {
                        currentIndex = leftChild(currentIndex)
                    } else {
                        currentIndex = rightChild(currentIndex)
                    }

                    if (currentIndex >= newTree.count) {
                        let sizeIncreaseNeeded: Int = currentIndex - newTree.count + 1
                        newTree.append(contentsOf: Array(repeating: "None", count: sizeIncreaseNeeded))
                    }

                    if (newTree[currentIndex] == "None") {
                        insertIndexFound = true
                    }
                }
                newTree[currentIndex] = c

                return BinarySearchTree.nodeTree(tree: newTree, count: count + 1)
        }
    }

    func contains(_ c: String) -> Bool {
        switch self {
            case .empty:
                return false
            case .nodeTree(let tree, _):
                return tree.contains(c)
        }
    }
}

File: kotlin/Exercises.kt
import java.io.BufferedReader
import java.io.FileReader
import java.io.IOException

fun change(amount: Long): Map<Int, Long> {
    require(amount >= 0) { "Amount cannot be negative" }
    
    val counts = mutableMapOf<Int, Long>()
    var remaining = amount
    for (denomination in listOf(25, 10, 5, 1)) {
        counts[denomination] = remaining / denomination
        remaining %= denomination
    }
    return counts
}

fun firstThenLowerCase(stringList: List<String>, filerFunction: (input: String) -> Boolean): String? {
    var outputString: String? = null
    var firstCaseFound = false
    for (string in stringList) {
        if (filerFunction(string) && !firstCaseFound) {
            outputString = string
            firstCaseFound = true
        }
    }

    return outputString?.lowercase()
}

class ChainableString(val phrase: String?) {
    fun and(stringSuffix: String): ChainableString {
        return ChainableString(phrase.plus(" ").plus(stringSuffix))
    }
}

fun say(string: String? = null): ChainableString {
    return ChainableString(string ?: "")
}

fun meaningfulLineCount(filePath: String): Long {
    var fileReader: FileReader

    try {
        fileReader = FileReader(filePath)
    } catch (e: Exception) {
        throw IOException("file")
    }

    val bufferedReader = BufferedReader(fileReader)
    val text: String = bufferedReader.use { reader ->
        reader.readText()
    }
    val lines = text.split("\n")

    var lineCount: Long = 0
    for (line in lines) {
        var firstChar = '#'
        var firstCharFound = false

        val chars = line.toCharArray()
        for (char in chars) {
            if (char != ' ' && char != '	' && char != ' ' && char.code != 13 && !firstCharFound) {
                firstChar = char
                firstCharFound = true
            }
        }

        if (firstChar != '#' && firstCharFound) {
            lineCount += 1
        }
    }
    return lineCount
}

class Quaternion(val a: Double, val b: Double, val c: Double, val d: Double) {
    companion object {
        val ZERO = Quaternion(0.0, 0.0, 0.0, 0.0)
        val I = Quaternion(0.0, 1.0, 0.0, 0.0)
        val J = Quaternion(0.0, 0.0, 1.0, 0.0)
        val K = Quaternion(0.0, 0.0, 0.0, 1.0)
    }

    operator fun times(a: Quaternion): Quaternion {
        return Quaternion((this.a * a.a) - (this.b * a.b) - (this.c * a.c) - (this.d * a.d),
                                       (this.a * a.b) + (this.b * a.a) + (this.c * a.d) - (this.d * a.c),
                                       (this.a * a.c) - (this.b * a.d) + (this.c * a.a) + (this.d * a.b),
                                       (this.a * a.d) + (this.b * a.c) - (this.c * a.b) + (this.d * a.a))
    }

    operator fun plus(a: Quaternion): Quaternion {
        return Quaternion(this.a + a.a, this.b + a.b, this.c + a.c, this.d + a.d)
    }

    override operator fun equals(a: Any?): Boolean {
        if (a is Quaternion) { return this.a == a.a && this.b == a.b && this.c == a.c && this.d == a.d } else { return false }
    }

    override fun toString(): String {
        var quatString = ""
        
        if (this.a != 0.0) { quatString = "${this.a}" }
        
        if (this.b == 1.0 && quatString.length == 0) { quatString = "${quatString}i" }
        else if (this.b == 1.0) { quatString = "${quatString}+i" }
        else if (this.b == -1.0) { quatString = "${quatString}-i" }
        else if (this.b > 0.0) { quatString = "${quatString}+${this.b}i" }
        else if (this.b < 0.0) { quatString = "${quatString}${this.b}i" }
        
        if (this.c == 1.0 && quatString.length == 0) { quatString = "${quatString}j" }
        else if (this.c == 1.0) { quatString = "${quatString}+j" }
        else if (this.c == -1.0) { quatString = "${quatString}-j" }
        else if (this.c > 0.0) { quatString = "${quatString}+${this.c}j" }
        else if (this.c < 0.0) { quatString = "${quatString}${this.c}j" }
        
        if (this.d == 1.0 && quatString.length == 0) { quatString = "${quatString}k" }
        else if (this.d == 1.0) { quatString = "${quatString}+k" }
        else if (this.d == -1.0) { quatString = "${quatString}-k" }
        else if (this.d > 0.0) { quatString = "${quatString}+${this.d}k" }
        else if (this.d < 0.0) { quatString = "${quatString}${this.d}k" }
        
        if (quatString.length == 0) { quatString = "0" }
        
        return quatString
    }

    fun coefficients(): List<Double> {
        return listOf(this.a, this.b, this.c, this.d)
    }

    fun conjugate(): Quaternion {
        return Quaternion(this.a, -this.b, -this.c, -this.d)
    }
}

sealed interface BinarySearchTree {
    fun insert(c: String): BinarySearchTree
    fun size(): Int
    fun contains(c: String): Boolean

    fun parent(i: Int): Int { return i/2 }
    fun leftChild(i: Int): Int { return i*2 }
    fun rightChild(i: Int): Int { return i*2 +1 }

    object Empty: BinarySearchTree {
        override fun insert(c: String): BinarySearchTree {
            val newTree = NodeTree(listOf("None", c), 1)
            return newTree
        }

        override fun size(): Int {
            return 0
        }

        override fun contains(c: String): Boolean {
            return false
        }
        
        override fun toString(): String {
            return "()"
        }
    }

    data class NodeTree(val tree: List<String>, val count: Int): BinarySearchTree {
        override fun insert(c: String): BinarySearchTree {
            var newTree: MutableList<String> = tree.toMutableList()
            var insertIndexFound: Boolean = false
            var currentIndex: Int = 1

            while (!insertIndexFound) {
                if (newTree[currentIndex].get(0).code > c.get(0).code) {
                    currentIndex = leftChild(currentIndex)
                } else {
                    currentIndex = rightChild(currentIndex)
                }

                if (currentIndex >= newTree.size) {
                    val sizeIncreaseNeeded: Int = currentIndex - newTree.size +1
                    newTree.addAll(List(sizeIncreaseNeeded) {"None"})
                }

                if (newTree[currentIndex] == "None") {
                    insertIndexFound = true
                }
            }
            newTree[currentIndex] = c

            return NodeTree(newTree.toList(), count +1)
        }

        override fun size(): Int {
            return count
        }

        override fun contains(c: String): Boolean {
            return tree.contains(c)
        }
        
        override fun toString(): String {
            var outputString: String = tree[1]
            if (tree.size > 2 && tree[2] != "None") {
                outputString = toStringHelper(outputString, 2)
            }
            if (tree.size > 3 && tree[3] != "None") {
                outputString = toStringHelper(outputString, 3)
            }
            return "(" + outputString + ")"
        }

        fun toStringHelper(c: String, i: Int): String {
            val parentStringIndex = c.indexOf(tree[parent(i)])
            var outputString: String = ""

            if (i == leftChild(parent(i))) {
                val insertingString = "(".plus(tree[i]).plus(")")
                outputString = c.substring(0, parentStringIndex).plus(insertingString).plus(c.substring(parentStringIndex))
            } else {
                val insertingString = "(".plus(tree[i]).plus(")")
                outputString = c.substring(0, parentStringIndex +1).plus(insertingString).plus(c.substring(parentStringIndex +1))
            }

            if (leftChild(i) < tree.size && tree[leftChild(i)] != "None") {
                outputString = toStringHelper(outputString, leftChild(i))
            }
            if (rightChild(i) < tree.size && tree[rightChild(i)] != "None") {
                outputString = toStringHelper(outputString, rightChild(i))
            }

            return outputString
        }
    }
}